From 0c9e54cda87601710683818107b63124b259b0ea Mon Sep 17 00:00:00 2001
From: Gwenhael Goavec-Merou <gwenhael@enjoy-digital.fr>
Date: Mon, 8 Apr 2024 11:39:27 +0200
Subject: [PATCH 2/3] sw/kernel/ethtool: missing wr-nic.h include

Signed-off-by: Gwenhael Goavec-Merou <gwenhael@enjoy-digital.fr>
---
 sw/kernel/ethtool.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/sw/kernel/ethtool.c b/sw/kernel/ethtool.c
index 122ce7d..0510fd2 100644
--- a/sw/kernel/ethtool.c
+++ b/sw/kernel/ethtool.c
@@ -17,6 +17,8 @@
 #include <linux/mii.h>
 #include <linux/ethtool.h>
 #include <linux/spinlock.h>
+
+#include "wr-nic.h"
 #ifdef ETHTOOL_SLINKSETTINGS
 
 /* return false if legacy contained non-0 deprecated fields
-- 
2.43.2

