From 6a1fe4c51b8a0894f7886228fc9a55af2b069842 Mon Sep 17 00:00:00 2001
From: Gwenhael Goavec-Merou <gwenhael@enjoy-digital.fr>
Date: Mon, 8 Apr 2024 12:11:53 +0200
Subject: [PATCH 2/2] kernel/fmc-fakedev.c: fix for kernel > 5.5.0

Signed-off-by: Gwenhael Goavec-Merou <gwenhael@enjoy-digital.fr>
---
 kernel/fmc-fakedev.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/kernel/fmc-fakedev.c b/kernel/fmc-fakedev.c
index ac27345..424c939 100644
--- a/kernel/fmc-fakedev.c
+++ b/kernel/fmc-fakedev.c
@@ -141,7 +141,11 @@ static void ff_work_fn(struct work_struct *work)
 
 	ff = ff_dev_create();
 	if (IS_ERR(ff)) {
+#if LINUX_VERSION_CODE < KERNEL_VERSION(5,5,0)
 		pr_warning("%s: can't re-create FMC devices\n", __func__);
+#else
+		pr_warn("%s: can't re-create FMC devices\n", __func__);
+#endif
 		return;
 	}
 	ret = fmc_device_register_n(ff->fmc, ff_nr_dev);
-- 
2.43.2

